packages/SwingSet/src/kernel/kernel.js

```diff
@@ -1121,22 +1154,44 @@ export default function buildKernel(
     }
 
     if (!kernelKeeper.isRunQueueEmpty()) {
-      return kernelKeeper.getNextMsg();
+      return kernelKeeper.getNextRunQueueMsg();
     }
     return undefined;
   }
 
+  function getNextAcceptanceMessage() {
+    if (!kernelKeeper.isAcceptanceQueueEmpty()) {
+      return kernelKeeper.getNextAcceptanceQueueMsg();
+    }
+    return undefined;
+  }
+
+  function maybeProcessNextMessage() {
+    /** @type {Promise<PolicyInput> | undefined} */
+    let resultPromise;
+    let message = getNextAcceptanceMessage();
+    if (message) {
```

Comment:
nit: I'd prefer a short-circuiting style:
```js
let message = getNextAcceptanceMessage();
if (message) {
  return processAcceptanceMessage(message);
}
message = getNextDeliveryMessage();
if (message) {
  return processAcceptanceMessage(message);
}
```

But the fact that you're returning a record, not just a promise, tells me that you've got something in mind for this, so ignore me if your plans don't align with my aesthetics.

BTW the return value of `getNextMessage` has an extra significance that needs to be kept in mind (it looks like your code does the right thing, but I wanted to be explicit about the requirement). If there's no work to be done, we don't commit any state changes. If e.g. `maybeProcessNextMessage` pulled something off a queue and *then* decided that we didn't need to do anything, we'd have a problem, because now there are outstanding state changes that need to be committed. If the kernel doesn't `commitCrank` before returning control to the host app, we're going to have consistency problems (sometimes the pop-from-queue changes get committed, sometimes they don't, depending upon what happens next time).

I think you're headed this way, but we should aim for a pattern in which crank commits happen at the end of every *crank*, regardless of whether that crank includes a delivery or not. When deliveries are unwound, we unwind the entire crank buffer, and we need to make sure that we still pop off whatever queue entry started the process. The `getNextMessage` call the `discardFailedDelivery` branch of `processQueueMessage` does this, and we need to make sure it still does the right thing no matter what queue it came from.

Reply:
> returning a record, not just a promise

In general I don't like have a return value of `Promise | undefined` as it would mean something is maybe async. I also prefer short circuiting, but in this case it didn't feel right.

I am planning on ripping this ugly function apart in the next PR, and didn't want to spend time making something that was right, but happy to revisit.

> I think you're headed this way, but we should aim for a pattern in which crank commits happen at the end of every _crank_, regardless of whether that crank includes a delivery or not

I believe that's already the case. The only time we don't create a crank is if there is no message at all. I will keep making sure that a crank is always created if a message is popped from a queue.